File: user-service/src/test/java/rs/raf/user_service/unit/UserControllerTest.java
package rs.raf.user_service.unit;

import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import rs.raf.user_service.controller.UserController;
import rs.raf.user_service.domain.dto.RoleRequestDto;
import rs.raf.user_service.domain.dto.UserDto;
import rs.raf.user_service.service.UserService;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

public class UserControllerTest {

    @Mock
    private UserService userService;

    @InjectMocks
    private UserController userController;

    public UserControllerTest() {
        MockitoAnnotations.openMocks(this);
    }

    @Test
    void testGetAllUsers() {
        Page<UserDto> users = new PageImpl<>(List.of(new UserDto()));
        when(userService.listUsers(any(PageRequest.class))).thenReturn(users);

        ResponseEntity<Page<UserDto>> response = userController.getAllUsers(0, 10);

        assertEquals(HttpStatus.OK, response.getStatusCode());
        assertEquals(users, response.getBody());
    }

    @Test
    void testGetUserRole() {
        when(userService.getUserRole(1L)).thenReturn("ROLE_ADMIN");

        ResponseEntity<String> response = userController.getUserRole(1L);

        assertEquals(HttpStatus.OK, response.getStatusCode());
        assertEquals("ROLE_ADMIN", response.getBody());
    }

    @Test
    void testAddRoleToUser_Success() {
        doNothing().when(userService).addRoleToUser(eq(1L), any(RoleRequestDto.class));

        RoleRequestDto roleRequestDto = new RoleRequestDto();
        ResponseEntity<Void> response = userController.addRoleToUser(1L, roleRequestDto);

        assertEquals(HttpStatus.OK, response.getStatusCode());
    }

    @Test
    void testAddRoleToUser_NotFound() {
        doThrow(new RuntimeException()).when(userService).addRoleToUser(eq(1L), any(RoleRequestDto.class));

        RoleRequestDto roleRequestDto = new RoleRequestDto();
        ResponseEntity<Void> response = userController.addRoleToUser(1L, roleRequestDto);

        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
    }

    @Test
    void testRemoveRoleFromUser_Success() {
        doNothing().when(userService).removeRoleFromUser(1L, 2L);

        ResponseEntity<Void> response = userController.removeRoleFromUser(1L, 2L);

        assertEquals(HttpStatus.OK, response.getStatusCode());
    }

    @Test
    void testRemoveRoleFromUser_NotFound() {
        doThrow(new RuntimeException()).when(userService).removeRoleFromUser(1L, 2L);

        ResponseEntity<Void> response = userController.removeRoleFromUser(1L, 2L);

        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
    }
}

File: user-service-kotlin/src/test/kotlin/pack/userservicekotlin/controller/UserControllerTest.kt
package pack.userservicekotlin.controller

import arrow.core.left
import arrow.core.right
import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.Mockito.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import org.springframework.http.MediaType
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.web.servlet.*
import pack.userservicekotlin.arrow.UserServiceError
import pack.userservicekotlin.domain.dto.external.UserResponseDto
import pack.userservicekotlin.domain.dto.role.RoleRequestDto
import pack.userservicekotlin.service.UserService
import pack.userservicekotlin.utils.JwtTokenUtil

@WebMvcTest(UserController::class)
@AutoConfigureMockMvc(addFilters = false)
class UserControllerTest {
    @Autowired lateinit var mockMvc: MockMvc

    @Autowired lateinit var objectMapper: ObjectMapper

    @MockitoBean
    lateinit var userService: UserService

    @MockitoBean
    lateinit var jwtTokenUtil: JwtTokenUtil

    @Test
    fun `getAllUsers should return page of users`() {
        val users = listOf(UserResponseDto(id = 1L, firstName = "John", lastName = "Doe", email = "j@example.com"))
        val page = PageImpl(users)

        `when`(userService.listUsers(PageRequest.of(0, 10))).thenReturn(page)

        mockMvc
            .get("/api/admin/users?page=0&size=10")
            .andExpect {
                status { isOk() }
                jsonPath("$.content[0].email") { value("j@example.com") }
            }
    }

    @Test
    fun `getUserRole should return role name`() {
        `when`(userService.getUserRole(1L)).thenReturn("AGENT".right())

        mockMvc
            .get("/api/admin/users/1/role")
            .andExpect {
                status { isOk() }
                content { string("AGENT") }
            }
    }

    @Test
    fun `getUserRole should return 404 if user not found`() {
        `when`(userService.getUserRole(1L)).thenReturn(UserServiceError.UserNotFound.left())

        mockMvc
            .get("/api/admin/users/1/role")
            .andExpect {
                status { isNotFound() }
                content { string("User not found") }
            }
    }

    @Test
    fun `addRoleToUser should return 200 when successful`() {
        val dto = RoleRequestDto(id = 1L)

        `when`(userService.addRoleToUser(Mockito.eq(1L), anyNonNull<RoleRequestDto>()))
            .thenReturn(Unit.right())

        mockMvc
            .post("/api/admin/users/1/role") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(dto)
            }.andExpect {
                status { isOk() }
            }
    }

    @Test
    fun `addRoleToUser should return 409 if role already assigned`() {
        val dto = RoleRequestDto(id = 1L)

        `when`(userService.addRoleToUser(Mockito.eq(1L), anyNonNull<RoleRequestDto>()))
            .thenReturn(UserServiceError.RoleAlreadyAssigned.left())

        mockMvc
            .post("/api/admin/users/1/role") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(dto)
            }.andExpect {
                status { isConflict() }
            }
    }

    @Test
    fun `removeRoleFromUser should return 200 on success`() {
        `when`(userService.removeRoleFromUser(1L, 1L)).thenReturn(Unit.right())

        mockMvc
            .delete("/api/admin/users/1/role/1")
            .andExpect {
                status { isOk() }
            }
    }

    @Test
    fun `removeRoleFromUser should return 400 if role not assigned`() {
        `when`(userService.removeRoleFromUser(1L, 1L)).thenReturn(UserServiceError.RoleNotAssigned.left())

        mockMvc
            .delete("/api/admin/users/1/role/1")
            .andExpect {
                status { isBadRequest() }
            }
    }

    // I have no idea what's this chatGPT thing
    inline fun <reified T> anyNonNull(): T = Mockito.any(T::class.java) ?: createInstance()

    inline fun <reified T : Any> createInstance(): T =
        when (T::class) {
            RoleRequestDto::class -> RoleRequestDto() as T
            else -> throw IllegalArgumentException("Provide default for ${T::class}")
        }
}
